flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -1252,41 +1285,48 @@ struct XEmboxOpConversion : public EmboxCommonConversion<fir::cg::XEmboxOp> {
           extent = rewriter.create<mlir::LLVM::SDivOp>(loc, i64Ty, extent,
                                                        operands[sliceOff + 2]);
         }
-        dest = insertExtent(rewriter, loc, dest, d, extent);
+        dest = insertExtent(rewriter, loc, dest, descIdx, extent);
 
         // store step (scaled by shaped extent)
         mlir::Value step = prevDim;
         if (hasSlice)
           step = rewriter.create<mlir::LLVM::MulOp>(loc, i64Ty, step,
                                                     operands[sliceOff + 2]);
-        dest = insertStride(rewriter, loc, dest, d, step);
-        d++;
+        dest = insertStride(rewriter, loc, dest, descIdx, step);
+        ++descIdx;
       }
 
       // compute the stride and offset for the next natural dimension
       prevDim =
           rewriter.create<mlir::LLVM::MulOp>(loc, i64Ty, prevDim, outerExtent);
-      prevPtrOff = rewriter.create<mlir::LLVM::MulOp>(loc, i64Ty, prevPtrOff,
-                                                      outerExtent);
+      if (constRows == 0)
+        prevPtrOff = rewriter.create<mlir::LLVM::MulOp>(loc, i64Ty, prevPtrOff,
+                                                        outerExtent);
 
       // increment iterators
-      shapeOff++;
+      ++shapeOff;
       if (hasShift)
-        shiftOff++;
+        ++shiftOff;
       if (hasSlice)
         sliceOff += 3;
     }
-    if (!xbox.subcomponent().empty()) {
-      // For each field in the path add the offset to base. In the most general
-      // case, some offsets must be computed since they are not be known until
-      // runtime.
-      TODO(xbox.getLoc(), "intra-entity slice in fir.embox codegen");
+    auto hasSubcomp = !xbox.subcomponent().empty();
+    if (hasSlice || hasSubcomp) {
+      llvm::SmallVector<mlir::Value> args = {base, ptrOffset};
+      args.append(gepArgs.rbegin(), gepArgs.rend());
```

Comment:
Does this works OK if the base type has a constant shape but element are dynamically sized ? e.g.: `!fir.array<5x6x7x!fir.char<1,?>>`

Reply:
It should. Unfortunately, there are some other bugs in lowering that need to be fixed as well.

Reply:
Right, sounds like it may be worth testing with FIR test cases then.